### ISSUE TITLE
CREATE INDEX ends in out of memory error

### ISSUE THREAD

**Issue by FabianErdm** [duckdb/duckdb#7760]
### What happens?

I've updated to the 0.8 version: Now I cannot create an index without an error:
CREATE INDEX i ON table (column)
Error: rapi_execute: Failed to run query
Error: Out of Memory Error: failed to allocate data of size 262KB (13.4GB/13.4GB used)

I am working within R, but the error also occurs when I use the console. In the 0.7.1 verison everything works.

I have tried PRAGMA memory_limit='32GB', but that results in the same error:
Error: rapi_execute: Failed to run query
Error: Out of Memory Error: failed to allocate data of size 262KB (31.9GB/32.0GB used)

### To Reproduce

CREATE INDEX i ON table (column)

### OS:

WIN 11

### DuckDB Version:

0.8

### DuckDB Client:

R / Console

### Full Name:

Erdmann

### Affiliation:

viadee

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by l1t1** [comment]
the size and the structure of your table?

**Comment by FabianErdm** [comment]
Size: ~86kk x 31, 30 varchar (also includes column to be indexed) and 1 timestamp.

table:
table_schema: main
table_type: BASE TABLE
is_typed: no

columns:
pk: all false

Let me know if you need any more information.

**Comment by l1t1** [comment]
86kk means 86000000 rows?

**Comment by l1t1** [comment]
you can replace 10 with 86000000 to test
```
D create table t31 as select now()+interval(i)second c,j,i::varchar c1 from range(10)t(i),range(30)s(j);
D from t31;
┌────────────────────────────┬───────┬─────────┐
│             c              │   j   │   c1    │
│  timestamp with time zone  │ int64 │ varchar │
├────────────────────────────┼───────┼─────────┤
│ 2023-06-01 12:03:21.364+00 │     0 │ 0       │
│ 2023-06-01 12:03:21.364+00 │     1 │ 0       │
│ 2023-06-01 12:03:21.364+00 │     2 │ 0       │
│ 2023-06-01 12:03:21.364+00 │     3 │ 0       │
│ 2023-06-01 12:03:21.364+00 │     4 │ 0       │
│ 2023-06-01 12:03:21.364+00 │     5 │ 0       │
│ 2023-06-01 12:03:21.364+00 │     6 │ 0       │
│ 2023-06-01 12:03:21.364+00 │     7 │ 0       │
│ 2023-06-01 12:03:21.364+00 │     8 │ 0       │
│ 2023-06-01 12:03:21.364+00 │     9 │ 0       │
│ 2023-06-01 12:03:21.364+00 │    10 │ 0       │
│ 2023-06-01 12:03:21.364+00 │    11 │ 0       │
│ 2023-06-01 12:03:21.364+00 │    12 │ 0       │
│ 2023-06-01 12:03:21.364+00 │    13 │ 0       │
│ 2023-06-01 12:03:21.364+00 │    14 │ 0       │
│ 2023-06-01 12:03:21.364+00 │    15 │ 0       │
│ 2023-06-01 12:03:21.364+00 │    16 │ 0       │
│ 2023-06-01 12:03:21.364+00 │    17 │ 0       │
│ 2023-06-01 12:03:21.364+00 │    18 │ 0       │
│ 2023-06-01 12:03:21.364+00 │    19 │ 0       │
│             ·              │     · │ ·       │
│             ·              │     · │ ·       │
│             ·              │     · │ ·       │
│ 2023-06-01 12:03:30.364+00 │    10 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    11 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    12 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    13 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    14 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    15 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    16 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    17 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    18 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    19 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    20 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    21 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    22 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    23 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    24 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    25 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    26 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    27 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    28 │ 9       │
│ 2023-06-01 12:03:30.364+00 │    29 │ 9       │
├────────────────────────────┴───────┴─────────┤
│ 300 rows (40 shown)                3 columns │
└──────────────────────────────────────────────┘
D create table z as pivot t31 on j using min(c1);
D from z;
┌──────────────────────┬─────────┬─────────┬─────────┬─────────┬───┬─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│          c           │    0    │    1    │   10    │   11    │ … │    4    │    5    │    6    │    7    │    8    │    9    │
│ timestamp with tim…  │ varchar │ varchar │ varchar │ varchar │   │ varchar │ varchar │ varchar │ varchar │ varchar │ varchar │
├──────────────────────┼─────────┼─────────┼─────────┼─────────┼───┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ 2023-06-01 12:03:2…  │ 0       │ 0       │ 0       │ 0       │ … │ 0       │ 0       │ 0       │ 0       │ 0       │ 0       │
│ 2023-06-01 12:03:2…  │ 1       │ 1       │ 1       │ 1       │ … │ 1       │ 1       │ 1       │ 1       │ 1       │ 1       │
│ 2023-06-01 12:03:2…  │ 2       │ 2       │ 2       │ 2       │ … │ 2       │ 2       │ 2       │ 2       │ 2       │ 2       │
│ 2023-06-01 12:03:2…  │ 3       │ 3       │ 3       │ 3       │ … │ 3       │ 3       │ 3       │ 3       │ 3       │ 3       │
│ 2023-06-01 12:03:2…  │ 4       │ 4       │ 4       │ 4       │ … │ 4       │ 4       │ 4       │ 4       │ 4       │ 4       │
│ 2023-06-01 12:03:2…  │ 5       │ 5       │ 5       │ 5       │ … │ 5       │ 5       │ 5       │ 5       │ 5       │ 5       │
│ 2023-06-01 12:03:2…  │ 6       │ 6       │ 6       │ 6       │ … │ 6       │ 6       │ 6       │ 6       │ 6       │ 6       │
│ 2023-06-01 12:03:2…  │ 7       │ 7       │ 7       │ 7       │ … │ 7       │ 7       │ 7       │ 7       │ 7       │ 7       │
│ 2023-06-01 12:03:2…  │ 8       │ 8       │ 8       │ 8       │ … │ 8       │ 8       │ 8       │ 8       │ 8       │ 8       │
│ 2023-06-01 12:03:3…  │ 9       │ 9       │ 9       │ 9       │ … │ 9       │ 9       │ 9       │ 9       │ 9       │ 9       │
├──────────────────────┴─────────┴─────────┴─────────┴─────────┴───┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┤
│ 10 rows                                                                                                31 columns (11 shown) │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
D create index on z("0");
Error: Not implemented Error: Index without a name not supported yet!
D create index idx on z("0");
D
```

**Comment by taniabogatsch** [comment]
There are a few known performance issues in some parts of the index code, and this issue is probably related. I almost have a PR up to fix that, I'll ping you once it is merged into master. Would be great if you could double-check then. :)

**Comment by FabianErdm** [comment]
Double check: indexing on 10 rows works (e.g. your example). Creating the example table is also not possible for me, as for 86kk rows duckdb crashes on my pc.
So it is indeed a performance issue, not a general issue (or issue with my general system)

**Comment by l1t1** [comment]
1000000 rows is ok
```
D create table t31 as select now()+interval(i)second c,j,i::varchar c1 from range(1e6::int)t(i),range(30)s(j);
100% ▕████████████████████████████████████████████████████████████▏
D .timer on
D create table z as pivot t31 on j using min(c1);
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 23.050 user 129.590031 sys 43.259077
D create index idx on z("0");
Run Time (s): real 0.532 user 2.121614 sys 0.873606
```

**Comment by l1t1** [comment]
1e7 rows failed in memory db
```
D create table t31 as select now()+interval(i)second c,j,i::varchar c1 from range(1e7::int)t(i),range(30)s(j);
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 85.844 user 82.462129 sys 3.354021
D create table z as pivot t31 on j using min(c1);
 28% ▕████████████████▊                                           ▏ Run Time (s): real 82.450 user 227.465058 sys 95.503812
Error: Out of Memory Error: could not allocate block of size 262KB (26.8GB/26.8GB used)
Database is launched in in-memory mode and no temporary directory is specified.
Unused blocks cannot be offloaded to disk.

Launch the database with a persistent storage back-end
Or set PRAGMA temp_directory='/path/to/tmp.tmp'
```

**Comment by l1t1** [comment]
use list+select to make the example table
```
D create table z1 as select c,list(c1)c2 from t31 group by c;
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 49.633 user 241.505148 sys 154.706192
D create table z2 as select c,c2[1],c2[2],c2[3],c2[4],c2[5],c2[6],c2[7],c2[8],c2[9],c2[10],c2[11],c2[12],c2[13],c2[14],c2[15],c2[16],c2[17],
19],c2[20],c2[21],c2[22],c2[23],c2[24],c2[25],c2[26],c2[27],c2[28],c2[29],c2[30] from z1;
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 6.776 user 40.497860 sys 37.143838
D
D create index idx on z2("c2[1]");
100% ▕████████████████████████████████████████████████████████████▏
Run Time (s): real 9.381 user 26.613771 sys 28.173781
D select count(*) from z2;
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│     10000000 │
└──────────────┘
Run Time (s): real 0.152 user 0.000000 sys 0.015600
```

**Comment by taniabogatsch** [comment]
Besides #8112, I am also working on decreasing the memory pressure of the ART during index creation (not the size that it takes up once created). I am leaving this comment to double-check once both PRs are on master.

**Comment by l1t1** [comment]
is this fixed in latest master?
